langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonWithType.java

```diff
@@ -0,0 +1,268 @@
+
+        TypeValuePair typeValuePair = new TypeValuePair(value, targetType);
+        BType sourceType = TypeChecker.getType(value);
+
+        if (unresolvedValues.contains(typeValuePair)) {
+            throw new BallerinaException(VALUE_LANG_LIB_CYCLIC_VALUE_REFERENCE_ERROR,
+                    BLangExceptionHelper.getErrorMessage(RuntimeErrors.CYCLIC_VALUE_REFERENCE, sourceType));
+        }
+
+        unresolvedValues.add(typeValuePair);
+
+        if (value == null) {
+            if (targetType.isNilable()) {
+                return null;
+            }
+            throw createError(VALUE_LANG_LIB_CONVERSION_ERROR,
+                    BLangExceptionHelper.getErrorMessage(RuntimeErrors.CANNOT_CONVERT_NIL, targetType));
+        }
+
+        List<BType> convertibleTypes = TypeConverter.getConvertibleTypesFromJson(value, targetType, new ArrayList<>());
+        if (convertibleTypes.isEmpty()) {
+            throw createConversionError(value, targetType);
+        } else if (convertibleTypes.size() > 1) {
+            throw createConversionError(value, targetType, AMBIGUOUS_TARGET);
+        }
+
+        BType matchingType = convertibleTypes.get(0);
+
+        Object newValue;
+        switch (sourceType.getTag()) {
+            case TypeTags.MAP_TAG:
+            case TypeTags.RECORD_TYPE_TAG:
+                newValue = convertMap((MapValue<?, ?>) value, matchingType, unresolvedValues, t, strand);
+                break;
+            case TypeTags.ARRAY_TAG:
+            case TypeTags.TUPLE_TAG:
+                newValue = convertArray((ArrayValue) value, matchingType, unresolvedValues, t, strand);
+                break;
+            case TypeTags.ERROR_TAG:
+                newValue = ((RefValue) value).copy(new HashMap<>());
+                break;
+            default:
+                if (((matchingType.getTag() >= TypeTags.XML_ELEMENT_TAG &&
+                        matchingType.getTag() <= TypeTags.XML_TEXT_TAG) || matchingType.getTag() == TypeTags.XML_TAG)) {
+                    try {
+                        newValue = XMLFactory.parse(((StringValue) value).getValue());
+                        break;
+                    } catch (Throwable e) {
+                        throw createConversionError(value, targetType, e.getMessage());
+                    }
+                }
+
+                // handle primitive values
+                if (sourceType.getTag() <= TypeTags.BOOLEAN_TAG) {
+                    if (TypeChecker.checkIsType(value, matchingType)) {
+                        newValue = value;
+                    } else {
+                        // Has to be a numeric conversion.
+                        newValue = TypeConverter.convertValues(matchingType, value);
+                    }
+                    break;
+                }
+                // should never reach here
+                throw BallerinaErrors.createConversionError(value, targetType);
+        }
+
+        unresolvedValues.remove(typeValuePair);
```

Comment:
Do we have to remove since we'll be creating a new list for each call to the method? Can you check others too.

Reply:
Ack

Reply:
But, There can be situations like
```ballerina
public function main() {
    json anObject = { name : "anObject", value : 10, sub : { subName : "subObject", subValue : 10 }};
    string k = anObject.toJsonString();
}
```

if we remove this line :`unresolvedValues.remove(typeValuePair);` ,  `10` is duplicated isn't it ?